File: apps/mobile/tool/dump_sync_wire_fixture.dart
import 'dart:convert';
import 'dart:io';

import 'package:naviwealth/core/sync/sync_api_client.dart';

void main(List<String> args) {
  if (args.length != 1) {
    stderr.writeln('usage: dump_sync_wire_fixture <fixture-name>');
    exitCode = 64;
    return;
  }

  final value = switch (args.single) {
    'sync_v2_client_push_row_change' => _clientPushChange().toJson(),
    'sync_v2_client_sync_request' => <String, Object?>{
      'device_id': 'device-a',
      'since': 41,
      'changes': <Object?>[_clientPushChange().toJson()],
    },
    final unknown => _unknown(unknown),
  };

  stdout.writeln(const JsonEncoder.withIndent('  ').convert(value));
}

RowChange _clientPushChange() => const RowChange(
  table: 'fin:accounts',
  id: 'acc-1',
  payload: <String, Object?>{
    'id': 'acc-1',
    'name': 'Cash',
    'hlc': '1716381000123.0000-device-a',
    'deleted_at': null,
  },
  version: '1716381000123.0000-device-a',
  deleted: false,
);

Never _unknown(String name) {
  stderr.writeln('unknown sync wire fixture: $name');
  exit(64);
}

File: apps/backend/src/bin/dump-sync-wire-fixture.rs
use naviwealth_backend::routes::sync::{RowAck, SyncResponse};
use naviwealth_backend::sync::store::RowChange;
use serde_json::json;
use std::collections::BTreeMap;

fn main() {
    let fixture = std::env::args().nth(1).unwrap_or_else(|| {
        eprintln!("usage: dump-sync-wire-fixture <fixture-name>");
        std::process::exit(64);
    });

    let output = match fixture.as_str() {
        "sync_v3_server_tombstone_row_change" => {
            serde_json::to_string_pretty(&server_tombstone()).unwrap()
        }
        "sync_v3_server_sync_response" => serde_json::to_string_pretty(&SyncResponse {
            seq: 42,
            changes: vec![server_tombstone()],
            more: false,
            accepted: vec![RowAck {
                table: "fin:accounts".into(),
                id: "acc-1".into(),
            }],
            domain_generations: BTreeMap::from([("finance".into(), 0), ("health".into(), 3)]),
        })
        .unwrap(),
        "sync_v3_server_empty_response" => serde_json::to_string_pretty(&SyncResponse {
            seq: 42,
            changes: vec![],
            more: false,
            accepted: vec![],
            domain_generations: BTreeMap::from([("finance".into(), 0)]),
        })
        .unwrap(),
        "sync_v3_server_more_response" => serde_json::to_string_pretty(&SyncResponse {
            seq: 42,
            changes: vec![server_tombstone()],
            more: true,
            accepted: vec![],
            domain_generations: BTreeMap::from([("health".into(), 3)]),
        })
        .unwrap(),
        "sync_v3_server_no_accepted_response" => serde_json::to_string_pretty(&SyncResponse {
            seq: 41,
            changes: vec![],
            more: false,
            accepted: vec![],
            domain_generations: BTreeMap::from([("finance".into(), 0)]),
        })
        .unwrap(),
        other => {
            eprintln!("unknown sync wire fixture: {other}");
            std::process::exit(64);
        }
    };

    println!("{output}");
}

fn server_tombstone() -> RowChange {
    RowChange {
        table: "health:health_metrics".into(),
        id: "metric-1".into(),
        payload: Some(json!({})),
        version: "1716381000124.0000-device-b".into(),
        deleted: true,
        device_id: "device-b".into(),
        seq: 42,
        generation: 3,
    }
}
